armory/utils/version.py

```diff
@@ -0,0 +1,64 @@
+"""
+ARMORY Versions use "Semantic Version" scheme where stable releases will have versions
+like `0.14.6`.  Armory uses the most recent git tag for versioning. For example if the
+most recent git tag is `v0.14.6`, then the version will be `0.14.6`.
+
+If you are a developer, the version will be constructed from the most recent tag plus a
+suffix of gHASH where HASH is the short hash of the most recent commit. For example,
+if the most recent git tag is v0.14.6 and the most recent commit hash is 1234567 then
+the version will be 0.14.6.g1234567. This scheme does differ from the scm strings
+which also have a commit count and date in them like 1.0.1.dev2+g0c5ffd9.d20220314181920
+which is a bit ungainly.
+"""
+
+
+import re
+
+from armory.logs import log
+
+
+def version_tuple(version_str: str) -> tuple:
+    return tuple(map(int, (version_str.split("."))))
+
+
+def trim_version(version_str: str = "0.0.0") -> str:
+    git_tag_regex = re.compile(r"(?P<version>[vV]?\d+(?:\.\d+){0,2})")
+    if (tag_match := git_tag_regex.match(version_str)) is not None:
+        return tag_match.group("version")
+    return version_str
+
+
+def get_metadata_version(package: str, version=None) -> str:
+    try:
+        from importlib.metadata import distribution, version, PackageNotFoundError
+    except ModuleNotFoundError:
+        from importlib_metadata import distribution, version, PackageNotFoundError
+
+    try:
+        if (version := distribution(package).version) is not None:
```

Comment:
The `:=` operator was introduced in python 3.8, iirc. Adding it here would break our code for python 3.7.

Reply:
Yep, fails for me in python 3.7:
```
>>> import armory
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david.slater/git/davidslater/armory/armory/__init__.py", line 6, in <module>
    from armory.utils import version, typedef
  File "/Users/david.slater/git/davidslater/armory/armory/utils/version.py", line 26
    if (tag_match := git_tag_regex.match(version_str)) is not None:
                  ^
SyntaxError: invalid syntax
```

Reply:
The `__about__.py` file is modified dynamically using a [build hook](https://github.com/christopherwoodall/armory/blob/976905143a50f9f128190b4ef5042e4a9bce74de/pyproject.toml#L141), so that anytime the package is installed- locally- the `VCS_VERSION` will update. That said, once the package is built- e.g. sdist/wheel- that version is then static.

Since the `.git` directory is not included in the distribution and is specifically ignored in the [`.dockerignore`](https://github.com/christopherwoodall/armory/blob/976905143a50f9f128190b4ef5042e4a9bce74de/.dockerignore#L2), once published, there is no other way to retrieve the release tag unless it is written somewhere.  

Thank you for catching the walrus ops, I bumped the Python version to 3.8 not realizing there was still code with specific support for 3.7.

Reply:
How does the versioning work with editable pip installs?

Reply:
> How does the versioning work with editable pip installs?

When building containers, the version is pulled when [the package is being built](https://github.com/christopherwoodall/armory/blob/develop/docker/build.py#L80); tag is pulled from the local `git` repository then added to the "metadata", and finally included in the sdist.

On a local machine, the tags are pulled directly from the `.git` metadata on every `pip install -e `. 


One minor change to the pip install is that instead of specifying a requirements file- e.g. `pip install -e host-requirements.txt`, the developer would instead run `pip install -e '.[developer,host]'` depending on what tools they need.